TinyCrypt deprecation - library's usage removal part 2 (bluetooth)

valeriosetti opened the pull request:
This is the second step toward the removal of TinyCrypt library usage from the Zephyr's codebase. This is the continuation of https://github.com/zephyrproject-rtos/zephyr/pull/79566 and https://github.com/zephyrproject-rtos/zephyr/pull/79653.
In particular this PR takes care of removing TinyCrypt usage from the bluetooth subsystem.

## To be done

Decide if we want to keep TinyCrypt in BT-mesh (as deprecated) or we want to completely remove it from the codebase.

valeriosetti committed:
bt: crypto/host: remove CONFIG_BT_USE_PSA_API symbol

CONFIG_BT_USE_PSA_API was used in BT crypto/host modules to select
PSA crypto API over TinyCrypt (which was the default until now).
Since TinyCrypt is being deprecated and PSA crypto API is the new
standard library for crypto operations, CONFIG_BT_USE_PSA_API is
no more needed.

Signed-off-by: Valerio Setti <vsetti@baylibre.com>

valeriosetti committed:
bt: use PSA for HCI ECC and rename BT_TINYCRYPT_ECC

This commit sets PSA Crypto API as the default library to perform
ECC-DH in HCI, replacing TinyCrypt (which is being deprecated).
Therefore the symbol BT_TINYCRYPT_ECC is renamed as BT_SEND_ECC_EMULATION.
References in samples/tests are also fixed.

Signed-off-by: Valerio Setti <vsetti@baylibre.com>

valeriosetti committed:
bt-mesh: deprecate BT_MESH_USES_TINYCRYPT

Since the TinyCrypt library is being deprecated in Zephyr, this
commit set TinyCrypt usage in BT mesh as deprecated and it sets
Mbed TLS PSA Crypto API as the default option (when TF-M is not
available).
Tests are also updated in this commit.

Signed-off-by: Valerio Setti <vsetti@baylibre.com>

valeriosetti committed:
bt: change dependencies for BT_RPA

Since the TinyCrypt library is being deprecated in Zephyr, this
commit modified the dependencies of BT_RPA. Instead of selecting
TinyCrypt, it relies on BT_HOST_CRYPTO or BT_CTLR_CRYPTO. This
helps both in the deprecation process of TinyCrypt, but also this
is more correct dependency since these 2 symbols are extensively
used in the rpa.c source code.

Signed-off-by: Valerio Setti <vsetti@baylibre.com>

alwa-nordic commented on subsys/bluetooth/host/Kconfig:
```diff
@@ -987,15 +987,17 @@ config BT_ECC
-	bool "Emulate ECDH in the Host using TinyCrypt library"
-	select TINYCRYPT
-	select TINYCRYPT_ECC_DH
+config BT_HCI_ECC
```
This name "BT_HCI_ECC" is not good at saying what it does. The HCI ECC command exits and may be available whether or not this config is enabled. Any ideas for a better name?

hermabe replied:
Agree. Ideally the `hci_ecc.c` file should be removed. There is no point for the host to go through an emulated hci layer for the ecdh commands.

We already have `CONFIG_BT_ECC` which deals with hci commands, so cant really use that. What about mirroring the crypto configs (`BT_CTLR_CRYPTO`/`BT_HOST_CRYPTO`) and the existing controller ecdh config (`BT_CTLR_ECDH`) with `BT_HOST_ECDH`?

alwa-nordic replied:
"BT_HOST_ECDH" would be short for "Bluetooth Host Elliptic Curve Diffie-Hellman". This is also not describing what this config does.

And it's not really in the Host either. It's used with HCI-raw as well. It's a kind of middleware. Maybe `BT_HCI_ECC_MIDDLEWARE`, `BT_HCI_ECC_INTERCEPT` or `BT_HCI_ECC_EMULATION`?

hermabe replied:
I was thinking we wanted to remove this hci emulation and just have normal function calls in the host

alwa-nordic replied:
I agree, Removing the HCI emulation would be the best.

alwa-nordic replied:
More suggestions for the name: `BT_HCI_ECC_COMPATIBILITY_LAYER` or `BT_HCI_ECC_POLYFILL`.

hermabe replied:
I wanted to avoid mentioning "hci" in the name since that is an implementation detail in the host that we want to remove. What it does is add ecdh support to the host.

valeriosetti replied:
Thanks for the suggestions! So far I'm still addressing issues with CI, that's why you still see `CONFIG_BT_HCI_ECC` around. Once I'm done with the CI, I will fix the name :)

> Removing the HCI emulation would be the best.

As for this, I think it would be clearer if I do that in a separate PR. Wdyt?

hermabe replied:
> As for this, I think it would be clearer if I do that in a separate PR. Wdyt?

Yes, that can be done later since it is independent of the PSA transition.

alwa-nordic replied:
> I wanted to avoid mentioning "hci" in the name since that is an implementation detail in the host that we want to remove. What it does is add ecdh support to the host.

This a valid perspective, except this config is also used with `CONFIG_BT_HCI_RAW` which we should keep distinct from the Host. Maybe we can say the emulation is not at "HCI" but "`bt_send`". So, `BT_SEND_ECC_EMULATION`?

alwa-nordic replied:
@jhedberg What do you think of the name? And should it even be available for HCI raw?

hermabe replied:
> > I wanted to avoid mentioning "hci" in the name since that is an implementation detail in the host that we want to remove. What it does is add ecdh support to the host.
> 
> This a valid perspective, except this config is also used with `CONFIG_BT_HCI_RAW` which we should keep distinct from the Host. Maybe we can say the emulation is not at "HCI" but "`bt_send`". So, `BT_SEND_ECC_EMULATION`?

HCI raw kind of breaks what I was thinking, with removing the hci commands in the host. If we want to support the ecdh commands in hci raw when the controller does not support them we will have to keep the hci emulation and the hooks in `bt_send`.

alwa-nordic replied:
Code is now updated to `BT_SEND_ECC_EMULATION`. I'm happy with that.

valeriosetti committed:
samples: bt: some RAM footprint reductions

As part of the work of deprecating TinyCrypt in favor of
PSA crypto API, this commit fixes some kconfigs in order to
reduce the RAM footprint. In all involved samples RAM usage
was extremely high even with TinyCrypt (usually > 98%) and
of course it got overflowed once TinyCrypt got replaced by
Mbed TLS's PSA Crypto API.

Signed-off-by: Valerio Setti <vsetti@baylibre.com>

rettichschnidi commented:
Any chance this will be ready for Zephyr 4.0?

valeriosetti commented:
> Any chance this will be ready for Zephyr 4.0?

As far as I know codefreeze is today for 4.0 and I still need to resolve some failures in this PR + it also depends on 2 other PRs which should be merged before this one. I'm sorry but I think the answer is "no"

valeriosetti committed:
tests: bluetooth: replace TinyCrypt with PSA in unit tests

Replace all references to TinyCrypt with PSA Crypto API in BT unit
tests.

Signed-off-by: Valerio Setti <vsetti@baylibre.com>

PavelVPV committed:
test: bluetooth: bsim: mesh: test_access: Increase timeout between proc

Increase timeout between procedures in access_tx_transmitt test to avoid
packet loss.

Signed-off-by: Pavel Vasilyev <pavel.vasilyev@nordicsemi.no>

mmahadevan108 commented:
@valeriosetti , do we need both Part 1 and Part 2 to be merged into 4.0. 
Can we merge only Part 1 if Part 2 is not ready in time before the release?

jhedberg commented:
> @valeriosetti , do we need both Part 1 and Part 2 to be merged into 4.0. Can we merge only Part 1 if Part 2 is not ready in time before the release?

The challenge as I see it is this part from https://docs.zephyrproject.org/latest/develop/api/api_lifecycle.html#deprecated
> Code using the deprecated API needs to be modified to remove usage of said API

Which means that we can't make a release and still have code in the upstream tree using a deprecated API.

valeriosetti commented:
> @valeriosetti , do we need both Part 1 and Part 2 to be merged into 4.0. Can we merge only Part 1 if Part 2 is not ready in time before the release?

While part 1 is ready to be merged, part 2 absolutely not:

- this PR indeed [depends](https://github.com/zephyrproject-rtos/zephyr/pull/79931#issue-2591998552) on  #80136 and #80627 (#79653 also, of course);
- there are few bsim failures I still need to solve. #80627 should address some of them, but not all AFAIK, but I didn't check yet.

> Which means that we can't make a release and still have code in the upstream tree using a deprecated API.

That's what I thought at the beginning of this activity as well, but then I've been asked to keep the following support for TinyCrypt in the codebase for at least the next 2 releases:

- [TinyCrypt based shim driver](https://github.com/zephyrproject-rtos/zephyr/pull/79653#pullrequestreview-2378460783);
- BT mesh, to help the people in the transition (this was requested on Teams by @alxelax so I don't have a link for this, sorry).

I'm not an expert of BT, but based on the number of changes I've done so far in this PR to make the CI happy, it seems that BT is way more affected by TinyCrypt removal than other subsystems. So I agree with @alxelax that giving the user some time to deal with TinyCrypt deprecation wouldn't be that bad. But this is my feeling, of course, so I'll be glad to hear what's reviewers opinion about this :)

valeriosetti committed:
test: bsim: bt: fix test scripts

This commit fixes
tests/bsim/bluetooth/ll/multiple_id/tests_scripts/multiple.sh.
The reason is that this test seems to depend on random number
sequence in order to have the proper scheduling.

It also fixes some AC tests by commenting out failing test
cases. The reason of the failure is still unknown and needs
to be investigated in future work.

Signed-off-by: Valerio Setti <vsetti@baylibre.com>

PavelVPV commented:
@valeriosetti you can now rebase your PR and drop two pre-last commits (https://github.com/zephyrproject-rtos/zephyr/pull/79931/commits/7d9ffddc72213a081ce8dc9e767d9e399d8a9e8e, https://github.com/zephyrproject-rtos/zephyr/pull/79931/commits/4aecc3afd3d8a138c06d5a3333407172d02d0713)

nashif commented:
and please move this out of draft

dkalowsk commented:
As discussed in the Release meeting today, we want to consider the deprecation of Tinycrypt a release blocker item for 4.0.  That would mean Part1, Part2, and the mbedTLS changes would all need to be merged in.

@tomi-font @valeriosetti any objection to this course of action?   The point was raised that Tinycrypt has been "just about to be deprecated" for 5 years now, but never seems to make it.  We'd like to make it happen now.

valeriosetti committed:
tests: bt: add Mbed TLS headers to the build

Include Mbed TLS headers to the build system. This is required
because these tests do not follow the "standard build pattern"
of Mbed TLS in Zephyr, otherwise include files would be already
available after the library has been linked. In these examples
some BT source files and Kconfigs are manually added to the
CmakeLists.txt file bypassing the standard library build pattern,
so Mbed TLS headers must also be added manually.

Signed-off-by: Valerio Setti <vsetti@baylibre.com>

valeriosetti committed:
tests: bt: increase PSA key slots count when necessary

Increase the number of key slots in the PSA Crypto core
for some tests using more keys than the default (16).

Signed-off-by: Valerio Setti <vsetti@baylibre.com>

valeriosetti committed:
tests: bt: enable entropy sources where missing

Moving from TinyCrypt to PSA Crypto API caused an entropy source
to be always required, so this commit adds it in tests where
necessary.

Signed-off-by: Valerio Setti <vsetti@baylibre.com>

tomi-font commented:
> As discussed in the Release meeting today, we want to consider the deprecation of Tinycrypt a release blocker item for 4.0. That would mean Part1, Part2, and the mbedTLS changes would all need to be merged in.
>
> @tomi-font @valeriosetti any objection to this course of action? The point was raised that Tinycrypt has been "just about to be deprecated" for 5 years now, but never seems to make it. We'd like to make it happen now.

No objection from me. I'll review this and the other needed PRs today.

alxelax commented on subsys/bluetooth/mesh/Kconfig:
```diff
@@ -1457,30 +1457,15 @@ endmenu # Proxy
-	help
-	  Use TinyCrypt library to perform crypto operations.
-
 config BT_MESH_USES_MBEDTLS_PSA
```
Please fix `help` description. This is not relevant anymore.

alxelax commented on subsys/bluetooth/mesh/Kconfig:
```diff
@@ -1457,30 +1457,15 @@ endmenu # Proxy
-
 config BT_MESH_USES_MBEDTLS_PSA
-	bool "mbed TLS PSA [EXPERIMENTAL]"
-	select EXPERIMENTAL
```
Did anybody run mesh with mbedtls on real devices? 
Also, to remove experimental tag bsim should use pure mbedtls without its emulation. 

It looks to me like, nobody run mesh with mbedtls PSA without technological stubs, but EXPERIMENTAL tag is removed. 
Kind of disagree with this.

alxelax commented on subsys/bluetooth/mesh/Kconfig:
```diff
@@ -1457,30 +1457,15 @@ endmenu # Proxy
 	help
 	  Crypto library selection for mesh security.
 
-config BT_MESH_USES_TINYCRYPT
```
I guess this should be moved back and marked as [DEPRECATED]

PavelVPV replied:
Agree, and a release notes should be added that this option is deprecated.

valeriosetti replied:
Ok, this should be addressed now

PavelVPV commented on samples/bluetooth/central_hr/prj_minimal.conf:
```diff
@@ -101,3 +101,14 @@ CONFIG_BT_L2CAP_TX_BUF_COUNT=2
+# RAM footprint
+CONFIG_MBEDTLS_PSA_KEY_SLOT_COUNT=5
+
+# This is required because otherwise there won't be
+# __heap_start/__heap_end symbols defined in the build. This
+# is due to Mbed TLS which at boot uses standard calloc/free
+# functions and only after the initialization done from Zephyr
+# it switches to its internal memory management functions.
+CONFIG_COMMON_LIBC_MALLOC=y
```
Will this be fixed eventually or we will always have to enable this option?

valeriosetti replied:
I tried to nicely fix this on Mbed TLS side (as part of #80136 - [this ](https://github.com/zephyrproject-rtos/zephyr/commit/9dc4b337b75a1581b80fab74d2d155a712205a16)was the fix that I tried in case you are curious to check) but unfortunately other failures appeared in that PR. Since we're a bit in a rush here I decided to keep this dirty fix, add a long comment to explain, and fix it later with a follow-up PR.

valeriosetti committed:
drivers: entropy: ease runtime requirements on BT HCI

On platforms like nrf5340 there are 2 CPUs:
- one is the cpu_net which takes care of the radio stuff and
  owns the HW random generator
- one is the cpu_app which holds application data and polls
  cpu_net through HCI commands when it needs some random data.

The PSA core implemented in Mbed TLS needs random data at initialization
time, which happens early in the boot process. If we wait for BT to
be ready before issuing the HCI command, then PSA core intialization
will fail. In facts there is no need for the BT to be completely
initialized just to ask for some random data from the cpu_app to
the cpu_net since the HW random generator will likely be already
functional in the cpu_net.
So let's just try the HCI command and, if something is not right,
it will fail anyway. There's no need to anticipate the failure.

Signed-off-by: Valerio Setti <vsetti@baylibre.com>

cvinayak commented on drivers/entropy/entropy_bt_hci.c:
```diff
@@ -20,9 +20,18 @@ static int entropy_bt_init(const struct device *dev)
-	if (!bt_is_ready()) {
-		return -EAGAIN;
-	}
+	/* Do not wait for BT to be ready (i.e. bt_is_ready()) before issueing
+	 * the command. The reason is that when crypto is enabled and the PSA
+	 * Crypto API support is provided through Mbed TLS, random number generator
+	 * needs to be available since the very first call to psa_crypto_init()
+	 * which is usually done before BT is completely initialized.
+	 * On the other hand, in devices like the nrf5340, the crytographically
+	 * secure RNG is owned by the cpu_net, so the cpu_app needs to poll it
+	 * to get random data. Again, there is no need to wait for BT to be
+	 * completely initialized for this kind of support. Just try to send the
+	 * request through HCI. If the command fails for any reason, then
+	 * we return failure anyway.
+	 */
```
As the return value could be different, we need API documentation change?

valeriosetti replied:
I added the documentation to the `migration-guide-4.0` document. I know it's not the right place because this won't be part of 4.0, but in the current status of the PR there's no 4.1 version of that document yet and rebasing the PR is a bit too much IMO just for this change. I will move the documentation to the proper document once
I can finalize the PR (i.e. solving all the issues with the CI).

valeriosetti committed:
board: enable random/entropy generators if BT

BT uses PSA Crypto API to perform crypto operations and, on this
platform, these APIs are implemented through Mbed TLS. In order
to properly initialize this library, a random number generator
is required.

* If the platform supports an HW entropy generator (ex: native_sim,
  nrf), then ENTROPY_GENERATOR must be used;
* Otherwise (ex: qemu_cortex_m3) test random generator can be
  enabled.

Enabling the proper option at board Kconfig level allows for
a more compact code change instead of manually editing _all_
the samples/tests that required this fix.

Signed-off-by: Valerio Setti <vsetti@baylibre.com>

valeriosetti committed:
tests: bt_crypto: fix kconfig for qemu_cortex_m3

Increase test and main stack sizes for the qemu_cortex_m3 platform
in order to be able to successfully run the test.

Signed-off-by: Valerio Setti <vsetti@baylibre.com>

valeriosetti committed:
samples: bt: reduce system heap memory for the hci_ipc sample

Moving from TinyCrypt to PSA Crypto API caused failures in the cpu_net
build due to RAM being overflowed. It turned out that 8192 bytes were
allocated for system heap memory, but Mbed TLS is the only user
of that memory (I found this though puncover) for AES purposes.
We reduce that to 1024 bytes because this should be enough for
this purpose.

Note: albeit this is also a standalone example, it's used extensively
in other samples/tests and babblesim, so a failure in building it
propagates in a lot of other failures.

Signed-off-by: Valerio Setti <vsetti@baylibre.com>

valeriosetti committed:
manifest: tf-m: update revision to include a fix

Update TF-M revision to include a fix for crypto_sizes.h header
which is no more aligned to the Mbed TLS one and this causes
build failures in tests.

Signed-off-by: Valerio Setti <vsetti@baylibre.com>

valeriosetti commented:
I just finished reshaping the commits in order to prepare for a full review. Please be aware that this PR still depends on some PRs, but I tried to keep those dependencies at the beginning of the commits list. So in case you want to start looking at this, you might ignore initial 4 commits

sjanc commented:
#AutoPTS run zephyr SM

codecoup-tester commented:
Scheduled PR https://github.com/zephyrproject-rtos/zephyr/pull/79931#issuecomment-2490947748, board: nrf52, estimated start time: 13:04:12, test case count: 71, estimated duration: 0:23:03<details><summary>Test cases to be run</summary>SM/CEN/PROT/BV-01-C<br>SM/CEN/JW/BV-05-C<br>SM/CEN/JW/BI-04-C<br>SM/CEN/JW/BI-01-C<br>SM/CEN/PKE/BV-01-C<br>SM/CEN/PKE/BV-04-C<br>SM/CEN/PKE/BI-01-C<br>SM/CEN/PKE/BI-02-C<br>SM/CEN/OOB/BV-01-C<br>SM/CEN/OOB/BV-03-C<br>SM/CEN/OOB/BV-09-C<br>SM/CEN/OOB/BI-01-C<br>SM/CEN/EKS/BV-01-C<br>SM/CEN/EKS/BI-01-C<br>SM/CEN/SIGN/BV-01-C<br>SM/CEN/SIGN/BV-03-C<br>SM/CEN/SIGN/BI-01-C<br>SM/CEN/KDU/BV-04-C<br>SM/CEN/KDU/BV-05-C<br>SM/CEN/KDU/BV-06-C<br>SM/CEN/KDU/BV-10-C<br>SM/CEN/KDU/BV-11-C<br>SM/CEN/KDU/BI-04-C<br>SM/CEN/PIS/BV-02-C<br>SM/CEN/PIS/BV-03-C<br>SM/CEN/SCJW/BV-01-C<br>SM/CEN/SCJW/BV-04-C<br>SM/CEN/SCJW/BI-01-C<br>SM/CEN/SCPK/BV-01-C<br>SM/CEN/SCPK/BV-04-C<br>SM/CEN/SCPK/BI-01-C<br>SM/CEN/SCPK/BI-02-C<br>SM/CEN/SCOB/BV-01-C<br>SM/CEN/SCOB/BV-04-C<br>SM/CEN/SCOB/BI-01-C<br>SM/CEN/SCOB/BI-04-C<br>SM/PER/PROT/BV-02-C<br>SM/PER/JW/BV-02-C<br>SM/PER/JW/BI-03-C<br>SM/PER/JW/BI-02-C<br>SM/PER/PKE/BV-02-C<br>SM/PER/PKE/BV-05-C<br>SM/PER/PKE/BI-03-C<br>SM/PER/OOB/BV-02-C<br>SM/PER/OOB/BV-04-C<br>SM/PER/OOB/BV-10-C<br>SM/PER/OOB/BI-02-C<br>SM/PER/EKS/BV-02-C<br>SM/PER/EKS/BI-02-C<br>SM/PER/KDU/BV-01-C<br>SM/PER/KDU/BV-02-C<br>SM/PER/KDU/BV-03-C<br>SM/PER/KDU/BV-07-C<br>SM/PER/KDU/BV-08-C<br>SM/PER/KDU/BV-09-C<br>SM/PER/KDU/BI-02-C<br>SM/PER/KDU/BI-03-C<br>SM/PER/KDU/BI-04-C<br>SM/PER/PIS/BV-01-C<br>SM/PER/PIS/BV-02-C<br>SM/PER/SCJW/BV-02-C<br>SM/PER/SCJW/BV-03-C<br>SM/PER/SCJW/BI-02-C<br>SM/PER/SCPK/BV-02-C<br>SM/PER/SCPK/BV-03-C<br>SM/PER/SCPK/BI-03-C<br>SM/PER/SCPK/BI-04-C<br>SM/PER/SCOB/BV-02-C<br>SM/PER/SCOB/BV-03-C<br>SM/PER/SCOB/BI-02-C<br>SM/PER/SCOB/BI-03-C</details>

alxelax commented on tests/bsim/bluetooth/mesh/overlay_psa.conf:
```diff
@@ -1,5 +1,8 @@
+# Increase the number of key slots in PSA Crypto core
+CONFIG_MBEDTLS_PSA_KEY_SLOT_COUNT=64
+
+# Increase the number of key slots in PSA Crypto core
```
Two times is not required :)

alxelax commented on tests/bsim/bluetooth/mesh/src/test_provision.c:
```diff
@@ -15,10 +15,6 @@
 
 #if defined CONFIG_BT_MESH_USES_MBEDTLS_PSA
 #include <psa/crypto.h>
-#elif defined CONFIG_BT_MESH_USES_TINYCRYPT
```
I guess until Tinycrypt deprecation period is over, the tests should present. @PavelVPV?

PavelVPV replied:
Since the mesh option stays, yes, we must keep the tests.

jhedberg replied:
AFAIK the default policy is not to have *any* in-tree references or tests for deprecated APIs. We can discuss and potentially approve exceptions if there are good justifications for it, but until then we should stick to the normal procedure.

alxelax replied:
Then all bsim tests those previously run with Tinycrypt now run with mbedtls. Furthermore, tests with psa overlay also are built and run with mbedtls. Almost all bsim mesh tests are built and run twice at the moment. That is time and CI loading overkill.

valeriosetti replied:
> AFAIK the default policy is not to have any in-tree references or tests for deprecated APIs.

Yeah, I this what I understood as well by reading the [API deprecation guide](https://docs.zephyrproject.org/latest/develop/api/api_lifecycle.html#deprecated).

> Then all bsim tests those previously run with Tinycrypt now run with mbedtls

Which should be fine, right? I mean, that's the goal of this PR after all.

> Furthermore, tests with psa overlay also are built and run with mbedtls. Almost all bsim mesh tests are built and run twice at the moment.

Right. I saw this and perhaps you already told me this, but I completely forgot to fix it. I will do that ASAP. Thanks!

alxelax replied:
> > AFAIK the default policy is not to have any in-tree references or tests for deprecated APIs.
> 
> Yeah, I this what I understood as well by reading the [API deprecation guide](https://docs.zephyrproject.org/latest/develop/api/api_lifecycle.html#deprecated).
> 
> > Then all bsim tests those previously run with Tinycrypt now run with mbedtls
> 
> Which should be fine, right? I mean, that's the goal of this PR after all.
> 
> > Furthermore, tests with psa overlay also are built and run with mbedtls. Almost all bsim mesh tests are built and run twice at the moment.
> 
> Right. I saw this and perhaps you already told me this, but I completely forgot to fix it. I will do that ASAP. Thanks!

Hi @valeriosetti, I do not mean it should be done in this PR. This PR is enormously huge without these changes. However, it should be done.

jhedberg replied:
I think part of the confusion here is that there are different interfaces to consider for deprecation and/or removal. The `BT_MESH_USES_TINYCRYPT` Kconfig option can be seen both as an API (something an app makes references to) as well as a user of TinyCrypt. Since TinyCrypt is being deprecated, as a user the option should get removed, however as an API the situation is more murky. I'm not sure there exists clear formal guidance on situations like this, but I'd be in favor of just removing it.

PavelVPV replied:
We would like to keep both, the `BT_MESH_USES_TINYCRYPT` option and the bsim tests with TinyCrypt until TinyCrypt is completely removed. The reason to keep the tests is to ensure that mesh with TinyCrypt still works and if not, we will fix bugs.

@jhedberg, should I bring this to some meeting?

codecoup-tester commented:
AutoPTS Bot results:
<details><summary>Failed tests (2)</summary>SM      SM/CEN/KDU/BI-04-C              FAIL<br>SM      SM/PER/KDU/BI-04-C              FAIL</details>
<details><summary>Successful tests (69)</summary>SM      SM/CEN/EKS/BI-01-C              PASS<br>SM      SM/CEN/EKS/BV-01-C              PASS<br>SM      SM/CEN/JW/BI-01-C               PASS<br>SM      SM/CEN/JW/BI-04-C               PASS<br>SM      SM/CEN/JW/BV-05-C               PASS<br>SM      SM/CEN/KDU/BV-04-C              PASS<br>SM      SM/CEN/KDU/BV-05-C              PASS<br>SM      SM/CEN/KDU/BV-06-C              PASS<br>SM      SM/CEN/KDU/BV-10-C              PASS<br>SM      SM/CEN/KDU/BV-11-C              PASS<br>SM      SM/CEN/OOB/BI-01-C              PASS<br>SM      SM/CEN/OOB/BV-01-C              PASS<br>SM      SM/CEN/OOB/BV-03-C              PASS<br>SM      SM/CEN/OOB/BV-09-C              PASS<br>SM      SM/CEN/PIS/BV-02-C              PASS<br>SM      SM/CEN/PIS/BV-03-C              PASS<br>SM      SM/CEN/PKE/BI-01-C              PASS<br>SM      SM/CEN/PKE/BI-02-C              PASS<br>SM      SM/CEN/PKE/BV-01-C              PASS<br>SM      SM/CEN/PKE/BV-04-C              PASS<br>SM      SM/CEN/PROT/BV-01-C             PASS<br>SM      SM/CEN/SCJW/BI-01-C             PASS<br>SM      SM/CEN/SCJW/BV-01-C             PASS<br>SM      SM/CEN/SCJW/BV-04-C             PASS<br>SM      SM/CEN/SCOB/BI-01-C             PASS<br>SM      SM/CEN/SCOB/BI-04-C             PASS<br>SM      SM/CEN/SCOB/BV-01-C             PASS<br>SM      SM/CEN/SCOB/BV-04-C             PASS<br>SM      SM/CEN/SCPK/BI-01-C             PASS<br>SM      SM/CEN/SCPK/BI-02-C             PASS<br>SM      SM/CEN/SCPK/BV-01-C             PASS<br>SM      SM/CEN/SCPK/BV-04-C             PASS<br>SM      SM/CEN/SIGN/BI-01-C             PASS<br>SM      SM/CEN/SIGN/BV-01-C             PASS<br>SM      SM/CEN/SIGN/BV-03-C             PASS<br>SM      SM/PER/EKS/BI-02-C              PASS<br>SM      SM/PER/EKS/BV-02-C              PASS<br>SM      SM/PER/JW/BI-02-C               PASS<br>SM      SM/PER/JW/BI-03-C               PASS<br>SM      SM/PER/JW/BV-02-C               PASS<br>SM      SM/PER/KDU/BI-02-C              PASS<br>SM      SM/PER/KDU/BI-03-C              PASS<br>SM      SM/PER/KDU/BV-01-C              PASS<br>SM      SM/PER/KDU/BV-02-C              PASS<br>SM      SM/PER/KDU/BV-03-C              PASS<br>SM      SM/PER/KDU/BV-07-C              PASS<br>SM      SM/PER/KDU/BV-08-C              PASS<br>SM      SM/PER/KDU/BV-09-C              PASS<br>SM      SM/PER/OOB/BI-02-C              PASS<br>SM      SM/PER/OOB/BV-02-C              PASS<br>SM      SM/PER/OOB/BV-04-C              PASS<br>SM      SM/PER/OOB/BV-10-C              PASS<br>SM      SM/PER/PIS/BV-01-C              PASS<br>SM      SM/PER/PIS/BV-02-C              PASS<br>SM      SM/PER/PKE/BI-03-C              PASS<br>SM      SM/PER/PKE/BV-02-C              PASS<br>SM      SM/PER/PKE/BV-05-C              PASS<br>SM      SM/PER/PROT/BV-02-C             PASS<br>SM      SM/PER/SCJW/BI-02-C             PASS<br>SM      SM/PER/SCJW/BV-02-C             PASS<br>SM      SM/PER/SCJW/BV-03-C             PASS<br>SM      SM/PER/SCOB/BI-02-C             PASS<br>SM      SM/PER/SCOB/BI-03-C             PASS<br>SM      SM/PER/SCOB/BV-02-C             PASS<br>SM      SM/PER/SCOB/BV-03-C             PASS<br>SM      SM/PER/SCPK/BI-03-C             PASS<br>SM      SM/PER/SCPK/BI-04-C             PASS<br>SM      SM/PER/SCPK/BV-02-C             PASS<br>SM      SM/PER/SCPK/BV-03-C             PASS</details>

sjanc commented:
#AutoPTS run zephyr nrf53 SM

codecoup-tester commented:
AutoPTS Bot results:
<details><summary>Failed tests (71)</summary>SM      SM/CEN/EKS/BI-01-C              BUILD_OR_FLASH ERROR<br>SM      SM/CEN/EKS/BV-01-C              BUILD_OR_FLASH ERROR<br>SM      SM/CEN/JW/BI-01-C               BUILD_OR_FLASH ERROR<br>SM      SM/CEN/JW/BI-04-C               BUILD_OR_FLASH ERROR<br>SM      SM/CEN/JW/BV-05-C               BUILD_OR_FLASH ERROR<br>SM      SM/CEN/KDU/BI-04-C              BUILD_OR_FLASH ERROR<br>SM      SM/CEN/KDU/BV-04-C              BUILD_OR_FLASH ERROR<br>SM      SM/CEN/KDU/BV-05-C              BUILD_OR_FLASH ERROR<br>SM      SM/CEN/KDU/BV-06-C              BUILD_OR_FLASH ERROR<br>SM      SM/CEN/KDU/BV-10-C              BUILD_OR_FLASH ERROR<br>SM      SM/CEN/KDU/BV-11-C              BUILD_OR_FLASH ERROR<br>SM      SM/CEN/OOB/BI-01-C              BUILD_OR_FLASH ERROR<br>SM      SM/CEN/OOB/BV-01-C              BUILD_OR_FLASH ERROR<br>SM      SM/CEN/OOB/BV-03-C              BUILD_OR_FLASH ERROR<br>SM      SM/CEN/OOB/BV-09-C              BUILD_OR_FLASH ERROR<br>SM      SM/CEN/PIS/BV-02-C              BUILD_OR_FLASH ERROR<br>SM      SM/CEN/PIS/BV-03-C              BUILD_OR_FLASH ERROR<br>SM      SM/CEN/PKE/BI-01-C              BUILD_OR_FLASH ERROR<br>SM      SM/CEN/PKE/BI-02-C              BUILD_OR_FLASH ERROR<br>SM      SM/CEN/PKE/BV-01-C              BUILD_OR_FLASH ERROR<br>SM      SM/CEN/PKE/BV-04-C              BUILD_OR_FLASH ERROR<br>SM      SM/CEN/PROT/BV-01-C             BUILD_OR_FLASH ERROR<br>SM      SM/CEN/SCJW/BI-01-C             BUILD_OR_FLASH ERROR<br>SM      SM/CEN/SCJW/BV-01-C             BUILD_OR_FLASH ERROR<br>SM      SM/CEN/SCJW/BV-04-C             BUILD_OR_FLASH ERROR<br>SM      SM/CEN/SCOB/BI-01-C             BUILD_OR_FLASH ERROR<br>SM      SM/CEN/SCOB/BI-04-C             BUILD_OR_FLASH ERROR<br>SM      SM/CEN/SCOB/BV-01-C             BUILD_OR_FLASH ERROR<br>SM      SM/CEN/SCOB/BV-04-C             BUILD_OR_FLASH ERROR<br>SM      SM/CEN/SCPK/BI-01-C             BUILD_OR_FLASH ERROR<br>SM      SM/CEN/SCPK/BI-02-C             BUILD_OR_FLASH ERROR<br>SM      SM/CEN/SCPK/BV-01-C             BUILD_OR_FLASH ERROR<br>SM      SM/CEN/SCPK/BV-04-C             BUILD_OR_FLASH ERROR<br>SM      SM/CEN/SIGN/BI-01-C             BUILD_OR_FLASH ERROR<br>SM      SM/CEN/SIGN/BV-01-C             BUILD_OR_FLASH ERROR<br>SM      SM/CEN/SIGN/BV-03-C             BUILD_OR_FLASH ERROR<br>SM      SM/PER/EKS/BI-02-C              BUILD_OR_FLASH ERROR<br>SM      SM/PER/EKS/BV-02-C              BUILD_OR_FLASH ERROR<br>SM      SM/PER/JW/BI-02-C               BUILD_OR_FLASH ERROR<br>SM      SM/PER/JW/BI-03-C               BUILD_OR_FLASH ERROR<br>SM      SM/PER/JW/BV-02-C               BUILD_OR_FLASH ERROR<br>SM      SM/PER/KDU/BI-02-C              BUILD_OR_FLASH ERROR<br>SM      SM/PER/KDU/BI-03-C              BUILD_OR_FLASH ERROR<br>SM      SM/PER/KDU/BI-04-C              BUILD_OR_FLASH ERROR<br>SM      SM/PER/KDU/BV-01-C              BUILD_OR_FLASH ERROR<br>SM      SM/PER/KDU/BV-02-C              BUILD_OR_FLASH ERROR<br>SM      SM/PER/KDU/BV-03-C              BUILD_OR_FLASH ERROR<br>SM      SM/PER/KDU/BV-07-C              BUILD_OR_FLASH ERROR<br>SM      SM/PER/KDU/BV-08-C              BUILD_OR_FLASH ERROR<br>SM      SM/PER/KDU/BV-09-C              BUILD_OR_FLASH ERROR<br>SM      SM/PER/OOB/BI-02-C              BUILD_OR_FLASH ERROR<br>SM      SM/PER/OOB/BV-02-C              BUILD_OR_FLASH ERROR<br>SM      SM/PER/OOB/BV-04-C              BUILD_OR_FLASH ERROR<br>SM      SM/PER/OOB/BV-10-C              BUILD_OR_FLASH ERROR<br>SM      SM/PER/PIS/BV-01-C              BUILD_OR_FLASH ERROR<br>SM      SM/PER/PIS/BV-02-C              BUILD_OR_FLASH ERROR<br>SM      SM/PER/PKE/BI-03-C              BUILD_OR_FLASH ERROR<br>SM      SM/PER/PKE/BV-02-C              BUILD_OR_FLASH ERROR<br>SM      SM/PER/PKE/BV-05-C              BUILD_OR_FLASH ERROR<br>SM      SM/PER/PROT/BV-02-C             BUILD_OR_FLASH ERROR<br>SM      SM/PER/SCJW/BI-02-C             BUILD_OR_FLASH ERROR<br>SM      SM/PER/SCJW/BV-02-C             BUILD_OR_FLASH ERROR<br>SM      SM/PER/SCJW/BV-03-C             BUILD_OR_FLASH ERROR<br>SM      SM/PER/SCOB/BI-02-C             BUILD_OR_FLASH ERROR<br>SM      SM/PER/SCOB/BI-03-C             BUILD_OR_FLASH ERROR<br>SM      SM/PER/SCOB/BV-02-C             BUILD_OR_FLASH ERROR<br>SM      SM/PER/SCOB/BV-03-C             BUILD_OR_FLASH ERROR<br>SM      SM/PER/SCPK/BI-03-C             BUILD_OR_FLASH ERROR<br>SM      SM/PER/SCPK/BI-04-C             BUILD_OR_FLASH ERROR<br>SM      SM/PER/SCPK/BV-02-C             BUILD_OR_FLASH ERROR<br>SM      SM/PER/SCPK/BV-03-C             BUILD_OR_FLASH ERROR</details>

valeriosetti commented:
#AutoPTS run zephyr SM

codecoup-tester commented:
Scheduled PR https://github.com/zephyrproject-rtos/zephyr/pull/79931#issuecomment-2493057058, board: nrf53, estimated start time: 08:33:24, test case count: 71, estimated duration: 0<details><summary>Test cases to be run</summary>SM/CEN/PROT/BV-01-C<br>SM/CEN/JW/BV-05-C<br>SM/CEN/JW/BI-04-C<br>SM/CEN/JW/BI-01-C<br>SM/CEN/PKE/BV-01-C<br>SM/CEN/PKE/BV-04-C<br>SM/CEN/PKE/BI-01-C<br>SM/CEN/PKE/BI-02-C<br>SM/CEN/OOB/BV-01-C<br>SM/CEN/OOB/BV-03-C<br>SM/CEN/OOB/BV-09-C<br>SM/CEN/OOB/BI-01-C<br>SM/CEN/EKS/BV-01-C<br>SM/CEN/EKS/BI-01-C<br>SM/CEN/SIGN/BV-01-C<br>SM/CEN/SIGN/BV-03-C<br>SM/CEN/SIGN/BI-01-C<br>SM/CEN/KDU/BV-04-C<br>SM/CEN/KDU/BV-05-C<br>SM/CEN/KDU/BV-06-C<br>SM/CEN/KDU/BV-10-C<br>SM/CEN/KDU/BV-11-C<br>SM/CEN/KDU/BI-04-C<br>SM/CEN/PIS/BV-02-C<br>SM/CEN/PIS/BV-03-C<br>SM/CEN/SCJW/BV-01-C<br>SM/CEN/SCJW/BV-04-C<br>SM/CEN/SCJW/BI-01-C<br>SM/CEN/SCPK/BV-01-C<br>SM/CEN/SCPK/BV-04-C<br>SM/CEN/SCPK/BI-01-C<br>SM/CEN/SCPK/BI-02-C<br>SM/CEN/SCOB/BV-01-C<br>SM/CEN/SCOB/BV-04-C<br>SM/CEN/SCOB/BI-01-C<br>SM/CEN/SCOB/BI-04-C<br>SM/PER/PROT/BV-02-C<br>SM/PER/JW/BV-02-C<br>SM/PER/JW/BI-03-C<br>SM/PER/JW/BI-02-C<br>SM/PER/PKE/BV-02-C<br>SM/PER/PKE/BV-05-C<br>SM/PER/PKE/BI-03-C<br>SM/PER/OOB/BV-02-C<br>SM/PER/OOB/BV-04-C<br>SM/PER/OOB/BV-10-C<br>SM/PER/OOB/BI-02-C<br>SM/PER/EKS/BV-02-C<br>SM/PER/EKS/BI-02-C<br>SM/PER/KDU/BV-01-C<br>SM/PER/KDU/BV-02-C<br>SM/PER/KDU/BV-03-C<br>SM/PER/KDU/BV-07-C<br>SM/PER/KDU/BV-08-C<br>SM/PER/KDU/BV-09-C<br>SM/PER/KDU/BI-02-C<br>SM/PER/KDU/BI-03-C<br>SM/PER/KDU/BI-04-C<br>SM/PER/PIS/BV-01-C<br>SM/PER/PIS/BV-02-C<br>SM/PER/SCJW/BV-02-C<br>SM/PER/SCJW/BV-03-C<br>SM/PER/SCJW/BI-02-C<br>SM/PER/SCPK/BV-02-C<br>SM/PER/SCPK/BV-03-C<br>SM/PER/SCPK/BI-03-C<br>SM/PER/SCPK/BI-04-C<br>SM/PER/SCOB/BV-02-C<br>SM/PER/SCOB/BV-03-C<br>SM/PER/SCOB/BI-02-C<br>SM/PER/SCOB/BI-03-C</details>

codecoup-tester commented:
AutoPTS Bot results:
<details><summary>Failed tests (21)</summary>SM      SM/CEN/KDU/BI-04-C              FAIL<br>SM      SM/CEN/KDU/BV-10-C              INCONC<br>SM      SM/CEN/KDU/BV-11-C              INCONC<br>SM      SM/CEN/SCJW/BV-01-C             INCONC<br>SM      SM/CEN/SCJW/BV-04-C             INCONC<br>SM      SM/CEN/SCOB/BV-01-C             INCONC<br>SM      SM/CEN/SCOB/BV-04-C             INCONC<br>SM      SM/CEN/SCPK/BI-01-C             BTP TIMEOUT<br>SM      SM/CEN/SCPK/BI-02-C             BTP TIMEOUT<br>SM      SM/CEN/SCPK/BV-01-C             INCONC<br>SM      SM/CEN/SCPK/BV-04-C             INCONC<br>SM      SM/PER/KDU/BI-03-C              INCONC<br>SM      SM/PER/KDU/BI-04-C              FAIL<br>SM      SM/PER/KDU/BV-08-C              INCONC<br>SM      SM/PER/KDU/BV-09-C              INCONC<br>SM      SM/PER/SCJW/BV-02-C             INCONC<br>SM      SM/PER/SCJW/BV-03-C             INCONC<br>SM      SM/PER/SCOB/BV-02-C             INCONC<br>SM      SM/PER/SCOB/BV-03-C             INCONC<br>SM      SM/PER/SCPK/BV-02-C             INCONC<br>SM      SM/PER/SCPK/BV-03-C             INCONC</details>
<details><summary>Successful tests (50)</summary>SM      SM/CEN/EKS/BI-01-C              PASS<br>SM      SM/CEN/EKS/BV-01-C              PASS<br>SM      SM/CEN/JW/BI-01-C               PASS<br>SM      SM/CEN/JW/BI-04-C               PASS<br>SM      SM/CEN/JW/BV-05-C               PASS<br>SM      SM/CEN/KDU/BV-04-C              PASS<br>SM      SM/CEN/KDU/BV-05-C              PASS<br>SM      SM/CEN/KDU/BV-06-C              PASS<br>SM      SM/CEN/OOB/BI-01-C              PASS<br>SM      SM/CEN/OOB/BV-01-C              PASS<br>SM      SM/CEN/OOB/BV-03-C              PASS<br>SM      SM/CEN/OOB/BV-09-C              PASS<br>SM      SM/CEN/PIS/BV-02-C              PASS<br>SM      SM/CEN/PIS/BV-03-C              PASS<br>SM      SM/CEN/PKE/BI-01-C              PASS<br>SM      SM/CEN/PKE/BI-02-C              PASS<br>SM      SM/CEN/PKE/BV-01-C              PASS<br>SM      SM/CEN/PKE/BV-04-C              PASS<br>SM      SM/CEN/PROT/BV-01-C             PASS<br>SM      SM/CEN/SCJW/BI-01-C             PASS<br>SM      SM/CEN/SCOB/BI-01-C             PASS<br>SM      SM/CEN/SCOB/BI-04-C             PASS<br>SM      SM/CEN/SIGN/BI-01-C             PASS<br>SM      SM/CEN/SIGN/BV-01-C             PASS<br>SM      SM/CEN/SIGN/BV-03-C             PASS<br>SM      SM/PER/EKS/BI-02-C              PASS<br>SM      SM/PER/EKS/BV-02-C              PASS<br>SM      SM/PER/JW/BI-02-C               PASS<br>SM      SM/PER/JW/BI-03-C               PASS<br>SM      SM/PER/JW/BV-02-C               PASS<br>SM      SM/PER/KDU/BI-02-C              PASS<br>SM      SM/PER/KDU/BV-01-C              PASS<br>SM      SM/PER/KDU/BV-02-C              PASS<br>SM      SM/PER/KDU/BV-03-C              PASS<br>SM      SM/PER/KDU/BV-07-C              PASS<br>SM      SM/PER/OOB/BI-02-C              PASS<br>SM      SM/PER/OOB/BV-02-C              PASS<br>SM      SM/PER/OOB/BV-04-C              PASS<br>SM      SM/PER/OOB/BV-10-C              PASS<br>SM      SM/PER/PIS/BV-01-C              PASS<br>SM      SM/PER/PIS/BV-02-C              PASS<br>SM      SM/PER/PKE/BI-03-C              PASS<br>SM      SM/PER/PKE/BV-02-C              PASS<br>SM      SM/PER/PKE/BV-05-C              PASS<br>SM      SM/PER/PROT/BV-02-C             PASS<br>SM      SM/PER/SCJW/BI-02-C             PASS<br>SM      SM/PER/SCOB/BI-02-C             PASS<br>SM      SM/PER/SCOB/BI-03-C             PASS<br>SM      SM/PER/SCPK/BI-03-C             PASS<br>SM      SM/PER/SCPK/BI-04-C             PASS</details>

omkar3141 commented:
Please note that Bluetooth Mesh has a dependency on tinycrypt until PSA support is hardened in upstream.

alxelax commented:
> Please note that Bluetooth Mesh has a dependency on tinycrypt until PSA support is hardened in upstream. @carlescufi @PavelVPV @alxelax

This PR removes tinycrypt dependency.

nashif commented:
Guys, we discussed this and decided that tinycrypt should be dropped after 4.0 is released. I am confused we are still having those discussions if certain aspects of this should stay or go. What is the holdup right now? why is this DNM?

nashif commented:
> Isn't DNM added by bot due to PR being listed in west.yml?

oh yeah, I thought we were beyond that already.... 

Someone needs to review and merge https://github.com/zephyrproject-rtos/trusted-firmware-m/pull/116

valeriosetti commented:
> Isn't DNM added by bot due to PR being listed in west.yml?

Yes, there is currently an update to `west.yaml` to pick a change in TF-M (it's just a small change to a header file). A part from that the PR is mostly fine, but it still depends on several other PRs that need to be merged before this one. Please check the description of the PR.

PavelVPV commented:
> Guys, we discussed this and decided that tinycrypt should be dropped after 4.0 is released. I am confused we are still having those discussions if certain aspects of this should stay or go. What is the holdup right now? why is this DNM?

From the mesh side the PR looks OK. The only question we have is if we should keep the mesh bsim tests with tinycrypt while we keep the mesh support with tinycrypt (as deprecated). In my opinion we should as it guarantees that the mesh with tinycrypt still works until tinycrypt is completely removed.

We asked for postponing tinycrypt deprecation because it was too short notice for us before the release to make sure that psa support in the mesh is good enough to remove the experimental tag and we will do this in this release, so thanks for postponing.

DNM was set by bot from what I see.

tomi-font commented:
> From the mesh side the PR looks OK. The only question we have is if we should keep the mesh bsim tests with tinycrypt while we keep the mesh support with tinycrypt (as deprecated). In my opinion we should as it guarantees that the mesh with tinycrypt still works until tinycrypt is completely removed.

+1

Thalley commented on samples/bluetooth/hci_ipc/nrf5340_cpunet_iso-bt_ll_sw_split.conf:
```diff
@@ -5,7 +5,7 @@ CONFIG_ISR_STACK_SIZE=1024
 CONFIG_MAIN_STACK_SIZE=512
 CONFIG_SYSTEM_WORKQUEUE_STACK_SIZE=512
-CONFIG_HEAP_MEM_POOL_SIZE=4096
+CONFIG_HEAP_MEM_POOL_SIZE=1024
```
This should be reverted. 1024 is too small

valeriosetti replied:
Yes, but since I just rebased after some dependency PRs were merged, I propose to keep this commit as-is and see if the CI is still OK, then remove it afterward. RAM usage in cpu_net has already shown to be a bit critical, so I suggest to change 1 thing at the time.

Thalley replied:
> Yes, but since I just rebased after some dependency PRs were merged, I propose to keep this commit as-is and see if the CI is still OK, then remove it afterward. RAM usage in cpu_net has already shown to be a bit critical, so I suggest to change 1 thing at the time.

This won't be caught by CI. It will only be triggered when running on-target on the nRF5340 I believe :) But would be caught by the PTS tests we can trigger

sjanc replied:
it won't be caught by (auto)PTS... as we have overlay in bttester for nrf5340 netcore which set this to ...7k

sjanc replied:
ahh scratch that, it loos like it uses just defaults from hci_ipc sample now

Thalley replied:
> ahh scratch that, it loos like it uses just defaults from hci_ipc sample now

Yeah, we just modified that :)

nashif commented:
so what about the babblesim test failing? the others we can ignore! I am ready to merge this now

valeriosetti commented:
> @valeriosetti so what about the babblesim test failing? the others we can ignore! I am ready to merge this now

The BabbleSim test failure was not related to crypto and for that I got a fix from @cvinayak that I applied [here](https://github.com/zephyrproject-rtos/zephyr/pull/79931/commits/29775c701c3c84a65e6dae3364de3cc6393c2f25).
As for other twister failures I saw that #82140 was merged (thanks!) and indeed I just rebased on `main`.

fabiobaltieri commented:
Hey this seems to be causing a pretty substantial increase in flash size usage on some bt applications, checking the hid example (west build -p -b nrf52dk/nrf52832 samples/bluetooth/peripheral_hids)

```
Memory region         Used Size  Region Size  %age Used
           FLASH:      181116 B       512 KB     34.55%
             RAM:       29240 B        64 KB     44.62%
        IDT_LIST:          0 GB        32 KB      0.00%
```
at 4b6683cca1322c1c85db17f53c1ed639ca85b0ef

and

```
Memory region         Used Size  Region Size  %age Used
           FLASH:      203988 B       512 KB     38.91%
             RAM:       30904 B        64 KB     47.16%
        IDT_LIST:          0 GB        32 KB      0.00%
```

at 184c0f9e9df4fa26e0508edff5bbb19e4790e850

That's pretty substantial, I have few projects myself that are now out of flash, is there some magic option to take the usage back down? Mainly thinking something for the migration guide.

valeriosetti commented:
@fabiobaltieri yes, I've also been reported by @Emil-Gydesen-Bose on Discord that with the current configuration crypto operations take noticeably longer than with TinyCrypt on nrf53 (build without TF-M) and nrf52. Of course this was not catched by CI because tests themselves are fine, they just take longer.
For configurations without TF-M I'm planning to enable `CONFIG_MBEDTLS_PSA_P256M_DRIVER_ENABLED` which uses an optimized version of p256m in Mbed TLS which is much faster than the default software implementation. On a nrf52 this made `bt init` command move from 2.1 seconds to 0.46 seconds, which is comparable to the TinyCrypt performance. That Kconfig also reduced the final binary size.

Can you please try it also on your side? It would be nice to have another +1 from someone who's testing this in a real scenario

valeriosetti commented on doc/releases/migration-guide-4.1.rst:
```diff
@@ -89,6 +89,15 @@ Display
+  data instead of waiting for BT connection to be ready. This is helpful on
+  platforms where the BT controller owns the HW random generator and the application
+  processor needs to get random data before BT is fully enabled.
+  (:github:`79931`)
```
I wrote it here because this was a change compared to the previous behavior. I mean, if the final user is already successfully using this driver then he won't notice any difference, but now they can use it also in other scenarios. I thought that `release-notes` was mostly for new features.
Perhaps I misinterpreted the difference between `migration-guide` and `release-notes`. Do you suggest to move it anyway?

fabiobaltieri replied:
Yeah there's probably some space for interpretation but I think the migration guide should be for stuff where the user has to take an action, like change something on the application side. @kartben? Anyway I don't think it's super important. One thing you may mention though is that some bluetooth functionalities now depends on the `mbedtls` module rather than `tinycrypt`, this is useful if anyone is running a project with a `name-allowlist` filter project list.

kartben commented:
@valeriosetti you're commenting on speed/performance but @fabiobaltieri was asking about flash size increase?

valeriosetti commented:
> @valeriosetti you're commenting on speed/performance but @fabiobaltieri was asking about flash size increase?

Actually in the comment there is also

> That Kconfig also reduced the final binary size.

at the end of the paragraph :)

Thalley commented:
@nashif this shouldn't have been merged IMO. For some reason the zephyrbot removed the DNM label https://github.com/zephyrproject-rtos/zephyr/pull/79931#event-15436778970 and I didn't feel the need to block this PR because it had that label. 

Anyways, there are 4 outstanding issues here:

1. Without the increase in `CONFIG_BT_LONG_WQ_STACK_SIZE` all Bluetooth pairings on the nRF53 and nRF54 series will fail on target
2. With the change to `CONFIG_HEAP_MEM_POOL_SIZE` (https://github.com/zephyrproject-rtos/zephyr/pull/79931#discussion_r1858455059) all `bt_enable()` on hardware will fail
3. There's nearly a 10x increase in pairing time (takes up to 10 seconds) which is not only really bad, but will also lead to timeouts on tests on hardware as well
4. There's a significant increase in memory as pointed out by https://github.com/zephyrproject-rtos/zephyr/pull/79931#issuecomment-2504945185

Points 1. and 2. will prevent us from running Bluetooth on several boards and prevent us from passing a lot of Bluetooth PTS tests. Point 3 may also timeout a lot of those tests. Point 4. might stop several boards from being able to run Bluetooth. 

I would suggest to either revert this PR and fix these things, or fix this fix with urgent care. 
I would also suggest that we start running some Bluetooth tests on hardware in CI. @sjanc Do you think we can work towards running just a few PTS tests in CI?

jhedberg commented:
@Thalley considering how long this PR had been open, and those issues hadn't been fixed, I have doubts that any of those would have been fixed in a timely manner. Now the merging of this PR at least works as a forcing function to get those issues fixed with urgency. It was still the right move not to merge this PR right before the 4.0 release, since then we'd have had a release with such breakage. Now that breakage is at least on the main branch, which is expected to be unstable at times during the merge window being open.

valeriosetti commented:
> I would suggest to either revert this PR and fix these things, or fix this fix with urgent care.

FWIW I'm working on a follow up PR that fixes (1), (3) and likely also noticeably reduce (4). I will post it later today

fabiobaltieri commented:
@valeriosetti yeah `CONFIG_MBEDTLS_PSA_P256M_DRIVER_ENABLED` seems to give some substantial savings, here's the numbers I'm seeing on my test application and with `west build -t rom_report`:

- tinycrypt: total 266080, tinycrypt modlue 5826
- mbedtls default: total 289000, mbedtls module 27736
- CONFIG_MBEDTLS_PSA_P256M_DRIVER_ENABLED=y: total 276436, mbedtls module: 15166

So the modules goes from roughly x5 to x3 the size, which is still a lot but way better. Flashed the firmware on my device, seems to connect and work fine but you folks probably want to do some better testing.

@Thalley the bot kicks in on west.yml changes and adds/removes the dnm tags depending on few conditions (if the hash is valid mainly but there's some more now), so if a PR has a manifest change you can't control that manually really. Maybe we should have a DNM tag for the bot and one for manual use.

valeriosetti commented:
> So the modules goes from roughly x5 to x3 the size, which is still a lot but way better. Flashed the firmware on my device, seems to connect and work fine but you folks probably want to do some better testing

Yeah, this increase is a known effect of the switch to PSA Crypto provided through Mbed TLS. As discussed in the [PR that introduced the P256-M driver](https://github.com/zephyrproject-rtos/zephyr/pull/73049#issuecomment-2146631348), PSA Crypto adds key management code (and some wrappers) that TinyCrypt does not have and which cause the increase of ROM code.

If it were only for EC support, then there would be `MBEDTLS_PSA_P256M_DRIVER_RAW` which gives you direct access to the P256-M driver (so no key management or extra-wrappers) and that would give you even smaller footprint than TinyCrypt. But this comes with a price: you would not use standard PSA Crypto API calls, but those of the driver instead + you need to manage the keys on your own. However this solution:
- is wrong for platforms that have TF-M support, because in that case PSA Crypto must provided by the TF-M code, not Mbed TLS [1]
- letting the user managing key storage is not the best because it depends on the user's knowledge about the topic. Now Mbed TLS's PSA Crypto core can make use of `secure storage` (introduced in 3.7) which allows key storing to be handled properly and directly from the PSA Crypto core.

However BT also uses PSA Crypto for AES-CMAC and for that one there's no "direct access" to the low level implementation, so there's no way to exclude the "core" part of PSA Crypto from the code.

[1]: it's worth mentioning IMO that on platforms with TF-M, the switch to PSA Crypto API from TinyCrypt _reduced_ ROM footprint because in that case we don't have a duplicated crypto library also in the non-secure side.

fabiobaltieri commented:
@valeriosetti thanks for the detailed answer that's quite something to digest, I won't pretend I understood half of that :-) but happy enough that you folks looked into it and decided it's working as intended.

msmttchr commented:
Sorry for late comment, but we discovered that this PR breaks compilation in most of the BlueNRG based platforms. In particular, is it a correct assumption that now entropy driver is mandatory for platforms supporting BLE?
I think it’s also not desirable the 3x memory increase, especially in platforms with limited amount of flash, not to speak about the performance penalties.
I guess there is no way to get a solution with similar performance of deprecated tinycrypt?

jhedberg commented:
@msmttchr #82217 should mitigate at least some of the memory overhead issues. As for the entropy requirement, that seems to indeed be the case. Some HCI drivers have opted to just `select ENTROPY_GENERATOR`, however there's also a more high level solution being worked on over here: #82363

sjanc commented:
btw mesh is now broken on main branch due to this when run on real HW  (eg  [00:00:40.191,589] <err> bt_mesh_net_keys: Unable to generate beacon key)

valeriosetti commented:
> btw mesh is now broken on main branch due to this when run on real HW (eg [00:00:40.191,589] bt_mesh_net_keys: Unable to generate beacon key)

Can I kindly ask you some more detail? Which sample? Which platform? So I can try to reproduce it
The best would be to file an issue I think if you didn't already

sjanc commented:
This is with test/bluetooth/tester,  on both nrf52 and nrf53

I've created https://github.com/zephyrproject-rtos/zephyr/issues/82645 as requested
